File: CastVideos-objc/Classes/MediaQueueViewController.m
// Copyright 2019 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "MediaQueueViewController.h"

#import <GoogleCast/GoogleCast.h>

#import "AppDelegate.h"
#import "MediaItem.h"

@interface MediaQueueViewController () <UITableViewDataSource,
                                        UITableViewDelegate,
                                        GCKSessionManagerListener,
                                        GCKRemoteMediaClientListener,
                                        GCKRequestDelegate> {
  NSTimer *_timer;

  // Queue
  IBOutlet UITableView *_tableView;

  // Queue/editing state.
  IBOutlet UIBarButtonItem *_editButton;
  BOOL _editing;

  GCKRemoteMediaClient *_mediaClient;
  GCKUIMediaController *_mediaController;
  GCKRequest *_queueRequest;
}

@end

@implementation MediaQueueViewController

- (void)viewDidLoad {
  NSLog(@"_tableView is %@", _tableView);
  _tableView.dataSource = self;
  _tableView.delegate = self;

  _editing = NO;

  GCKSessionManager *sessionManager = [GCKCastContext sharedInstance].sessionManager;
  [sessionManager addListener:self];
  if (sessionManager.hasConnectedCastSession) {
    [self attachToCastSession:sessionManager.currentCastSession];
  }

  UILongPressGestureRecognizer *recognizer =
      [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(handleLongPress:)];
  recognizer.minimumPressDuration = 2.0;  // 2 seconds
  [_tableView addGestureRecognizer:recognizer];
  _tableView.separatorColor = [UIColor clearColor];

  [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated {
  appDelegate.castControlBarsEnabled = NO;
  [super viewWillAppear:animated];
}

- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];

  _queueRequest = nil;
  _tableView.userInteractionEnabled = YES;

  if ([_mediaClient.mediaStatus queueItemCount] == 0) {
    [_editButton setEnabled:NO];
  } else {
    [_editButton setEnabled:YES];
  }

  [_tableView reloadData];
}

- (void)viewWillDisappear:(BOOL)animated {
  [super viewWillDisappear:animated];
}

#pragma mark - UI Actions

- (IBAction)toggleEditing:(id)sender {
  if (_editing) {
    _editButton.title = @"Edit";
    [_tableView setEditing:NO animated:YES];
    _editing = NO;
    if ([_mediaClient.mediaStatus queueItemCount] == 0) {
      [_editButton setEnabled:NO];
    }
  } else {
    _editButton.title = @"Done";
    [_tableView setEditing:YES animated:YES];
    _editing = YES;
  }
}

- (void)showErrorMessage:(NSString *)message {
  UIAlertView *alert = [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"Error", nil)
                                                  message:message
                                                 delegate:nil
                                        cancelButtonTitle:NSLocalizedString(@"OK", nil)
                                        otherButtonTitles:nil];
  [alert show];
}

- (void)handleLongPress:(UILongPressGestureRecognizer *)gestureRecognizer {
  CGPoint point = [gestureRecognizer locationInView:_tableView];
  NSIndexPath *indexPath = [_tableView indexPathForRowAtPoint:point];
  if (indexPath) {
    GCKMediaQueueItem *item = [_mediaClient.mediaStatus queueItemAtIndex:indexPath.row];
    if (item) {
      [self startRequest:[_mediaClient queueJumpToItemWithID:item.itemID]];
    }
  }
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  if (!_mediaClient || !_mediaClient.mediaStatus) {
    return 0;
  }

  return _mediaClient.mediaStatus.queueItemCount;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MediaCell"];

  GCKMediaQueueItem *item = [_mediaClient.mediaStatus queueItemAtIndex:indexPath.row];

  NSString *title = [item.mediaInformation.metadata stringForKey:kGCKMetadataKeyTitle];
  NSString *artist = [item.mediaInformation.metadata stringForKey:kGCKMetadataKeyArtist];
  if (!artist) {
    NSString *value = [item.mediaInformation.metadata stringForKey:kGCKMetadataKeyStudio];
    artist = value ? value : @"";
  }

  NSString *duration = (item.mediaInformation.streamDuration == INFINITY) ? @"" : [GCKUIUtils timeIntervalAsString:item.mediaInformation.streamDuration];
  NSString *detail = [NSString
      stringWithFormat:@"(%@) %@",
                       duration,
                       artist];
  UILabel *mediaTitle = (UILabel *)[cell viewWithTag:1];
  mediaTitle.text = title;

  UILabel *mediaOwner = (UILabel *)[cell viewWithTag:2];
  mediaOwner.text = detail;

  if (_mediaClient.mediaStatus.currentItemID == item.itemID) {
    cell.backgroundColor = [UIColor colorWithRed:15.0 / 255
                                           green:153.0 / 255
                                            blue:242.0 / 255
                                           alpha:0.1];
  } else {
    cell.backgroundColor = nil;
  }

  UIImageView *imageView = (UIImageView *)[cell.contentView viewWithTag:3];

  NSArray *images = item.mediaInformation.metadata.images;
  if (images && images.count > 0) {
    GCKImage *image = images[0];

    [[GCKCastContext sharedInstance].imageCache fetchImageForURL:image.URL
                                                      completion:^(UIImage *image) {
                                                        imageView.image = image;
                                                        [cell setNeedsLayout];
                                                      }];
  }

  return cell;
}

#pragma mark - UITableViewDelegate

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
  return YES;
}

- (void)tableView:(UITableView *)tableView
    moveRowAtIndexPath:(NSIndexPath *)sourceIndexPath
           toIndexPath:(NSIndexPath *)destinationIndexPath {
  if (sourceIndexPath.row == destinationIndexPath.row) {
    return;
  }
  GCKMediaQueueItem *sourceItem = [_mediaClient.mediaStatus queueItemAtIndex:sourceIndexPath.row];
  NSUInteger insertBeforeID = kGCKMediaQueueInvalidItemID;

  if (destinationIndexPath.row < (NSInteger)[_mediaClient.mediaStatus queueItemCount] - 1) {
    GCKMediaQueueItem *beforeItem =
        [_mediaClient.mediaStatus queueItemAtIndex:destinationIndexPath.row];
    insertBeforeID = beforeItem.itemID;
  }

  [self startRequest:[_mediaClient queueMoveItemWithID:sourceItem.itemID
                                      beforeItemWithID:insertBeforeID]];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView
           editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
  return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView
    commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
     forRowAtIndexPath:(NSIndexPath *)indexPath {
  if (editingStyle == UITableViewCellEditingStyleDelete) {
    // Delete row.
    GCKMediaQueueItem *item = [_mediaClient.mediaStatus queueItemAtIndex:indexPath.row];
    if (item) {
      [self startRequest:[_mediaClient queueRemoveItemWithID:item.itemID]];
    }
  }
}

- (void)tableView:(UITableView *)tableView didEndEditingRowAtIndexPath:(NSIndexPath *)indexPath {
  // No-op.
}

#pragma mark - Session handling

- (void)attachToCastSession:(GCKCastSession *)castSession {
  _mediaClient = castSession.remoteMediaClient;
  [_mediaClient addListener:self];
  [_tableView reloadData];
}

- (void)detachFromCastSession {
  [_mediaClient removeListener:self];
  _mediaClient = nil;
  [_tableView reloadData];
}

#pragma mark - GCKSessionManagerListener

- (void)sessionManager:(GCKSessionManager *)sessionManager
    didStartCastSession:(GCKCastSession *)session {
  [self attachToCastSession:session];
}

- (void)sessionManager:(GCKSessionManager *)sessionManager
    didSuspendCastSession:(GCKCastSession *)session
               withReason:(GCKConnectionSuspendReason)reason {
  [self detachFromCastSession];
}

- (void)sessionManager:(GCKSessionManager *)sessionManager
    didResumeCastSession:(GCKCastSession *)session {
  [self attachToCastSession:session];
}

- (void)sessionManager:(GCKSessionManager *)sessionManager
    willEndCastSession:(GCKCastSession *)session {
  [self detachFromCastSession];
}

#pragma mark - GCKRemoteMediaClientListener

- (void)remoteMediaClient:(GCKRemoteMediaClient *)client
     didUpdateMediaStatus:(GCKMediaStatus *)mediaStatus {
  [_tableView reloadData];
}

- (void)remoteMediaClientDidUpdateQueue:(GCKRemoteMediaClient *)client {
  [_tableView reloadData];
}

#pragma mark - Request scheduling

- (void)startRequest:(GCKRequest *)request {
  _queueRequest = request;
  _queueRequest.delegate = self;
  _tableView.userInteractionEnabled = NO;
}

#pragma mark - GCKRequestDelegate

- (void)requestDidComplete:(GCKRequest *)request {
  if (request == _queueRequest) {
    _queueRequest = nil;
    _tableView.userInteractionEnabled = YES;
  }
}

- (void)request:(GCKRequest *)request didFailWithError:(GCKError *)error {
  if (request == _queueRequest) {
    _queueRequest = nil;
    _tableView.userInteractionEnabled = YES;
    [self showErrorMessage:[NSString
                               stringWithFormat:@"Queue request failed:\n%@", error.description]];
  }
}

- (void)requestWasReplaced:(GCKRequest *)request {
  if (request == _queueRequest) {
    _queueRequest = nil;
    _tableView.userInteractionEnabled = YES;
  }
}

@end

File: CastVideos-swift/Classes/MediaQueueViewController.swift
// Copyright 2018 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import GoogleCast
import UIKit

@objc(MediaQueueViewController)
class MediaQueueViewController: UIViewController, UITableViewDataSource, UITableViewDelegate,
  GCKSessionManagerListener, GCKRemoteMediaClientListener, GCKRequestDelegate {
  private var timer: Timer!
  // Queue
  @IBOutlet private var _tableView: UITableView!
  // Queue/editing state.
  @IBOutlet private var _editButton: UIBarButtonItem!
  private var _editing = false
  private var mediaClient: GCKRemoteMediaClient!
  private var mediaController: GCKUIMediaController!
  private var queueRequest: GCKRequest!

  override func viewDidLoad() {
    print("_tableView is \(_tableView.description)")
    _tableView.dataSource = self
    _tableView.delegate = self
    _editing = false
    let sessionManager = GCKCastContext.sharedInstance().sessionManager
    sessionManager.add(self)
    if sessionManager.hasConnectedCastSession() {
      attach(to: sessionManager.currentCastSession!)
    }
    let recognizer = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPress))
    recognizer.minimumPressDuration = 2.0
    // 2 seconds
    _tableView.addGestureRecognizer(recognizer)
    _tableView.separatorColor = UIColor.clear
    super.viewDidLoad()
  }

  override func viewWillAppear(_ animated: Bool) {
    appDelegate?.isCastControlBarsEnabled = false
    super.viewWillAppear(animated)
  }

  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    queueRequest = nil
    _tableView.isUserInteractionEnabled = true
    if mediaClient.mediaStatus?.queueItemCount == 0 {
      _editButton.isEnabled = false
    } else {
      _editButton.isEnabled = true
    }
    _tableView.reloadData()
  }

  override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
  }

  // MARK: - UI Actions

  @IBAction func toggleEditing(_: Any) {
    if _editing {
      _editButton.title = "Edit"
      _tableView.setEditing(false, animated: true)
      _editing = false
      if mediaClient.mediaStatus?.queueItemCount == 0 {
        _editButton.isEnabled = false
      }
    } else {
      _editButton.title = "Done"
      _tableView.setEditing(true, animated: true)
      _editing = true
    }
  }

  func showErrorMessage(_ message: String) {
    let alert = UIAlertView(title: NSLocalizedString("Error", comment: ""),
                            message: message,
                            delegate: nil,
                            cancelButtonTitle: NSLocalizedString("OK", comment: ""),
                            otherButtonTitles: "")
    alert.show()
  }

  @objc func handleLongPress(_ gestureRecognizer: UILongPressGestureRecognizer) {
    let point: CGPoint = gestureRecognizer.location(in: _tableView)
    if let indexPath = _tableView.indexPathForRow(at: point) {
      let item: GCKMediaQueueItem? = mediaClient.mediaStatus?.queueItem(at: UInt(indexPath.row))
      if item != nil {
        start(mediaClient.queueJumpToItem(withID: (item?.itemID)!))
      }
    }
  }

  // MARK: - UITableViewDataSource

  func numberOfSections(in _: UITableView) -> Int {
    return 1
  }

  func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
    if (mediaClient == nil) || (mediaClient.mediaStatus == nil) {
      return 0
    }
    return Int(mediaClient.mediaStatus!.queueItemCount)
  }

  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "MediaCell")
    let item: GCKMediaQueueItem? = mediaClient.mediaStatus?.queueItem(at: UInt(indexPath.row))
    let title: String? = item?.mediaInformation.metadata?.string(forKey: kGCKMetadataKeyTitle)
    var artist: String? = item?.mediaInformation.metadata?.string(forKey: kGCKMetadataKeyArtist)
    if artist == nil {
      artist = item?.mediaInformation.metadata?.string(forKey: kGCKMetadataKeyStudio)
    }
    let detail: String? = "(\(GCKUIUtils.timeInterval(asString: (item?.mediaInformation.streamDuration)!))) \(artist ?? "")"
    let mediaTitle: UILabel? = (cell?.viewWithTag(1) as? UILabel)
    mediaTitle?.text = title
    let mediaOwner: UILabel? = (cell?.viewWithTag(2) as? UILabel)
    mediaOwner?.text = detail
    if mediaClient.mediaStatus?.currentItemID == item?.itemID {
      cell?.backgroundColor = UIColor(red: CGFloat(15.0 / 255), green: CGFloat(153.0 / 255),
                                      blue: CGFloat(242.0 / 255), alpha: CGFloat(0.1))
    } else {
      cell?.backgroundColor = nil
    }
    let imageView = (cell?.contentView.viewWithTag(3) as? UIImageView)
    if let images = item?.mediaInformation.metadata?.images(), images.count > 0 {
      let image = images[0] as? GCKImage
      GCKCastContext.sharedInstance().imageCache?.fetchImage(for: (image?.url)!,
                                                             completion: { (_ image: UIImage?) -> Void in
                                                               imageView?.image = image
                                                               cell?.setNeedsLayout()
      })
    }
    return cell!
  }

  // MARK: - UITableViewDelegate

  func tableView(_: UITableView, canMoveRowAt _: IndexPath) -> Bool {
    return true
  }

  func tableView(_: UITableView, moveRowAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {
    if sourceIndexPath.row == destinationIndexPath.row {
      return
    }
    let sourceItem = mediaClient.mediaStatus?.queueItem(at: UInt(sourceIndexPath.row))
    var insertBeforeID = kGCKMediaQueueInvalidItemID
    if destinationIndexPath.row < Int((mediaClient.mediaStatus?.queueItemCount)!) - 1 {
      let beforeItem: GCKMediaQueueItem? = mediaClient.mediaStatus?.queueItem(at: UInt(destinationIndexPath.row))
      insertBeforeID = (beforeItem?.itemID)!
    }
    start(mediaClient.queueMoveItem(withID: (sourceItem?.itemID)!, beforeItemWithID: insertBeforeID))
  }

  func tableView(_: UITableView, editingStyleForRowAt _: IndexPath) -> UITableViewCell.EditingStyle {
    return .delete
  }

  func tableView(_: UITableView, commit editingStyle: UITableViewCell.EditingStyle,
                 forRowAt indexPath: IndexPath) {
    if editingStyle == .delete {
      // Delete row.
      let item: GCKMediaQueueItem? = mediaClient.mediaStatus?.queueItem(at: UInt(indexPath.row))
      if item != nil {
        start(mediaClient.queueRemoveItem(withID: (item?.itemID)!))
      }
    }
  }

  func tableView(_: UITableView, didEndEditingRowAt _: IndexPath?) {}

  // MARK: - Session handling

  func attach(to castSession: GCKCastSession) {
    mediaClient = castSession.remoteMediaClient
    mediaClient.add(self)
    _tableView.reloadData()
  }

  func detachFromCastSession() {
    mediaClient.remove(self)
    mediaClient = nil
    _tableView.reloadData()
  }

  // MARK: - GCKSessionManagerListener

  func sessionManager(_: GCKSessionManager, didStart session: GCKCastSession) {
    attach(to: session)
  }

  func sessionManager(_: GCKSessionManager, didSuspend _: GCKCastSession,
                      with _: GCKConnectionSuspendReason) {
    detachFromCastSession()
  }

  func sessionManager(_: GCKSessionManager, didResumeCastSession session: GCKCastSession) {
    attach(to: session)
  }

  func sessionManager(_: GCKSessionManager, willEnd _: GCKCastSession) {
    detachFromCastSession()
  }

  // MARK: - GCKRemoteMediaClientListener

  func remoteMediaClient(_: GCKRemoteMediaClient, didUpdate _: GCKMediaStatus?) {
    _tableView.reloadData()
  }

  func remoteMediaClientDidUpdateQueue(_: GCKRemoteMediaClient) {
    _tableView.reloadData()
  }

  // MARK: - Request scheduling

  func start(_ request: GCKRequest) {
    queueRequest = request
    queueRequest.delegate = self
    _tableView.isUserInteractionEnabled = false
  }

  // MARK: - GCKRequestDelegate

  func requestDidComplete(_ request: GCKRequest) {
    if request == queueRequest {
      queueRequest = nil
      _tableView.isUserInteractionEnabled = true
    }
  }

  func request(_ request: GCKRequest, didFailWithError error: GCKError) {
    if request == queueRequest {
      queueRequest = nil
      _tableView.isUserInteractionEnabled = true
      showErrorMessage("Queue request failed:\n\(error.description)")
    }
  }

  func requestWasReplaced(_ request: GCKRequest) {
    if request == queueRequest {
      queueRequest = nil
      _tableView.isUserInteractionEnabled = true
    }
  }
}
